src/control/cmd/agent/infocache_test.go

```diff
@@ -0,0 +1,294 @@
+				fmt.Sprintf("expected: %s, got %s", tc.deviceName, response[0].GetInterface()))
+
+			// verifies that the neighbor response is what was expected
+			common.AssertTrue(t, response[tc.numDevices+1].GetInterface() == tc.neighbor,
+				fmt.Sprintf("expected: %s, got %s", tc.neighbor, response[tc.numDevices+1].GetInterface()))
+		})
+	}
+}
+
+// getResponse is called as a concurrent routine to access the info cache.
+// The cache contains responses with interface names specific to each NUMA node.
+// Aside from actually generating a failure due to a race condition, the response
+// data is checked to make sure it contains a device known to be associated with the
+// given numa node.
+func getResponse(t *testing.T, aiCache *attachInfoCache, numaNode int, wg *sync.WaitGroup) {
+	defer wg.Done()
+	res, err := aiCache.getResponse(numaNode)
+	common.AssertEqual(t, err, nil, "TestInfoCacheConcurrentAccess getResponse error")
+
+	resp := &mgmtpb.GetAttachInfoResp{}
+	if err = proto.Unmarshal(res, resp); err != nil {
+		t.Errorf("Expected error on proto.Unmarshal, got %+v", err)
+	}
+
+	deviceName := fmt.Sprintf("_node%d", numaNode)
+	if !strings.HasSuffix(resp.GetInterface(), deviceName) {
+		t.Errorf("TestInfoCacheConcurrentAccess response mismatch.  Devicename: %s, does not have suffix: %s", resp.GetInterface(), deviceName)
+	}
+}
+
+// TestInfoCacheConcurrentAccess launches maxConcurrent go routines
+// that concurrently access the info cache.  This test verifies that there are
+// no race conditions accessing the shared cache.
+func TestInfoCacheConcurrentAccess(t *testing.T) {
+	log, buf := logging.NewTestLogger(t.Name())
+	defer common.ShowBufferOnFailure(t, buf)
+	enabled := atm.NewBool(true)
+	scanResults := []netdetect.FabricScan{
+		{Provider: "ofi+sockets", DeviceName: "eth0_node0", NUMANode: 0, Priority: 0},
+		{Provider: "ofi+sockets", DeviceName: "eth1_node0", NUMANode: 0, Priority: 1},
+		{Provider: "ofi+sockets", DeviceName: "eth2_node0", NUMANode: 0, Priority: 2},
+		{Provider: "ofi+sockets", DeviceName: "eth3_node0", NUMANode: 0, Priority: 3},
+		{Provider: "ofi+sockets", DeviceName: "eth0_node1", NUMANode: 1, Priority: 0},
+		{Provider: "ofi+sockets", DeviceName: "eth1_node1", NUMANode: 1, Priority: 1},
+		{Provider: "ofi+sockets", DeviceName: "eth0_node2", NUMANode: 2, Priority: 0},
+		{Provider: "ofi+sockets", DeviceName: "eth1_node2", NUMANode: 2, Priority: 1},
+		{Provider: "ofi+sockets", DeviceName: "eth2_node2", NUMANode: 2, Priority: 2},
+		{Provider: "ofi+sockets", DeviceName: "eth3_node2", NUMANode: 2, Priority: 3}}
+
+	aiCache := attachInfoCache{log: log, enabled: enabled}
+	err := aiCache.initResponseCache(&mgmtpb.GetAttachInfoResp{}, scanResults)
+	common.AssertEqual(t, err, nil, "initResponseCache error")
+
+	var wg sync.WaitGroup
+	maxNumaNodes := 3
+	for i := 0; i < maxConcurrent; i++ {
+		wg.Add(1)
+		getResponse(t, &aiCache, i%maxNumaNodes, &wg)
+	}
+	wg.Wait()
```

Comment:
Unless I'm missing something, there's no concurrency here; you've just got a for loop with some fancy garland. There needs to be a `go foo()` in order to start a goroutine with a function.

I recommend something like this:
```go
	rand.Seed(time.Now().UnixNano()) // seed RNG
	for i := 0; i < maxConcurrent; i++ {
		wg.Add(1)
		go func() {
			time.Sleep(rand.Intn(100) * time.Microsecond)
			getResponse(t, &aiCache, i%maxNumaNodes, &wg)
		}()
	}
```

Reply:
Heh.  In my mind I had a version that called this as a go routine.  I must have fallen asleep at the wheel.  I'll go fix it.